src/app/clusters/fan-control-server/fan-control-server.cpp

```diff
@@ -0,0 +1,198 @@
+            }
+        }
+
+        // Adjust SpeedSetting from a percent value change for PercentSetting
+        // speed = ceil( SpeedMax * (percent * 0.01) )
+        uint8_t speedMax;
+        status = SpeedMax::Get(attributePath.mEndpointId, &speedMax);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status, ChipLogError(Zcl, "Failed to get SpeedMax with error: 0x%02x", status));
+
+        uint8_t currentSpeedSetting;
+        status = SpeedSetting::Get(attributePath.mEndpointId, &currentSpeedSetting);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                       ChipLogError(Zcl, "Failed to get SpeedSetting with error: 0x%02x", status));
+
+        float precent        = percentSetting;
+        uint8_t speedSetting = static_cast<uint8_t>(ceil(speedMax * (precent * 0.01)));
+
+        if (speedSetting != currentSpeedSetting)
+        {
+            status = SpeedSetting::Set(attributePath.mEndpointId, speedSetting);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to set SpeedSetting with error: 0x%02x", status));
+        }
+        break;
+    }
+    case SpeedSetting::Id: {
+        uint8_t speedSetting;
+        EmberAfStatus status = SpeedSetting::Get(attributePath.mEndpointId, &speedSetting);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status);
+
+        // If SpeedSetting is set to 0, the server SHALL set the FanMode attribute value to Off.
+        if (speedSetting == 0)
+        {
+            FanModeType currentFanMode;
+            status = FanMode::Get(attributePath.mEndpointId, &currentFanMode);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to get FanMode with error: 0x%02x", status));
+
+            if (currentFanMode != FanModeType::kOff)
+            {
+                status = FanMode::Set(attributePath.mEndpointId, FanModeType::kOff);
+                VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                               ChipLogError(Zcl, "Failed to write FanMode with error: 0x%02x", status));
+            }
+        }
+
+        // Adjust PercentSetting from a speed value change for SpeedSetting
+        // percent = floor( speed/SpeedMax * 100 )
+        uint8_t speedMax;
+        status = SpeedMax::Get(attributePath.mEndpointId, &speedMax);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status, ChipLogError(Zcl, "Failed to get SpeedMax with error: 0x%02x", status));
+
+        uint8_t currentPercentSetting;
+        status = PercentSetting::Get(attributePath.mEndpointId, &currentPercentSetting);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                       ChipLogError(Zcl, "Failed to get PercentSetting with error: 0x%02x", status));
+
+        float speed            = speedSetting;
+        uint8_t percentSetting = static_cast<uint8_t>(speed / speedMax * 100);
```

Comment:
Can this just be:
```suggestion
        uint8_t percentSetting = static_cast<uint8_t>(speed * 100 / speedMax);
```
or does that give different answers?